train.py

```diff
@@ -1,115 +1,167 @@
+
+        # NOTE: optional fields that may be missing from json but specified in config
+        optionalEntries = ["max acceleration", "max deceleration"]
+
+        usedFields = set()
+        config.pop('id')
+
+        for entry in config:
+
+            if config[entry] is None and entry in data:  # None -> no constraint
+
+                data.pop(entry)
+
+                usedFields.add(entry)
+
+            else:
+
+                if not isinstance(config[entry], dict) or config[entry].keys() != {'unit', 'value'}:
+
+                    raise ValueError("Configuration field '{}' should be specified as a dictionary with 'unit' and 'value' keys!".format(entry))
+
+                if entry in data or entry in optionalEntries:
+
+                    data[entry] = config[entry]
+
+                    usedFields.add(entry)
 
-            raise ValueError("Unknown train specified!")
+        if set(config) != usedFields:
 
-        # train mass parameters
+            raise ValueError("Redundant fields in train configuration: {}!".format(', '.join(set(config) - usedFields)))
 
-        self.mass = data['mass']  # train mass [kg]
+        # read data
+        self.mass = convertUnit(data['mass']['value'], data['mass']['unit'])  # train mass [kg]
 
-        self.g = data['g']  # acceleration of gravity [m/s^2]
+        self.rho = convertUnit(data['rho']['value'], data['rho']['unit'])  # rotating-mass factor [-]
 
-        self.rho = data['rho']  # rotating-mass factor [-]
+        if self.rho < 1:
 
-        # velocity limits
+            self.rho += 1  # 6% -> 0.06 -> 1.06
 
-        self.velocityMax = data['velocityMax']  # maximum train speed [m/s]
+        self.velocityMax = convertUnit(data['max speed']['value'], data['max speed']['unit'])  # maximum train speed [m/s]
 
-        # force limits
+        self.forceMax = convertUnit(data['max traction force']['value'], data['max traction force']['unit'])  # maximum traction force [N]
 
-        self.forceMax = data['forceMax']  # maximum traction force [N]
+        self.forceMin = convertUnit(-abs(data['max reg braking force']['value']), data['max reg braking force']['unit'])  # maximum regenerative braking force [N]
 
-        self.forceMin = data['forceMin']  # maximum regenerative braking force [N]
+        self.forceMinPn = convertUnit(-abs(data['max pn braking force']['value']), data['max pn braking force']['unit'])  # maximum pneumatic braking force [N]
 
-        self.forceMinPn = data['forceMinPn']  # maximum pneumatic braking force [N]
+        self.powerMax = convertUnit(data['max traction power']['value'], data['max traction power']['unit'])  # maximum traction power [W]
 
-        # power limits and losses
+        self.powerMin = convertUnit(-abs(data['max reg braking power']['value']), data['max reg braking power']['unit'])  # maximum regenerative braking power [W]
 
-        self.powerMax = data['powerMax']  # maximum traction power [W]
+        if 'max acceleration' in data:
```

Comment:
If not specified, rather than dealing with Nones later in the code, we could assign here the default value infinity.

Reply:
I have the same convention on power, force and acceleration. None means unconstrained. In the json file, a null is also interpreted as a None. When the field is not there in the json file it is the same as defining it null.